### PR TITLE
Ensure we PUT and POST non-collection resources with the correct x-if…

### PR DESCRIPTION
**Pull request by mhammond**
…-unmodified-since header

Even though we don't attempt to overwrite these resources (ie, we only upload them when we think they are new), there's still a risk we will race with another client even in that case.

### CODE REVIEW

File: components/sync15/src/client.rs [hunk 1]
```diff
@@ -25,16 +29,39 @@ pub struct Sync15StorageClientInit {
     pub tokenserver_url: Url,
 }
 
+fn get_response_timestamp(resp: &mut Response) -> error::Result<ServerTimestamp> {
```
Comment (thomcc) [inline components/sync15/src/client.rs]: Does this need &mut? (either way it won't after the network patch so it doesn't matter much).

File: components/sync15/src/client.rs [hunk 2]
```diff
@@ -25,16 +29,39 @@ pub struct Sync15StorageClientInit {
...
+        .and_then(|v| v.to_str().ok())
+        .and_then(|s| ServerTimestamp::from_str(s).ok())
+        .ok_or_else(|| ErrorKind::MissingServerTimestamp)?)
+}
+
+/// A TimestampedResponse is used as a wrapper for any non-collection response
+/// we may want to update with an x-if-unmodified-since header to ensure we are
+/// overwriting what we think we are.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct TimestampedResponse<T: DeserializeOwned> {
```
Comment (thomcc) [inline components/sync15/src/client.rs]: Do we need `T: DeserializeOwned` here too?

File: components/sync15/src/client.rs (hunk 1 above)
Comment (mhammond) [reply components/sync15/src/client.rs]: ah yeah, thanks. `.json` needs mut, but this doesn't.

**Comment by mhammond** [comment]
Because this is persisted, I need to make sure nothing breaks when we drop that persisted state.

**Comment by thomcc** [comment]
We turned on commit signing, and can't land this until it gets signed FYI.

File: components/sync15/src/client.rs [hunk 3]
```diff
@@ -25,16 +27,38 @@ pub struct Sync15StorageClientInit {
...
+        .get(X_LAST_MODIFIED)
+        .and_then(|v| v.to_str().ok())
+        .and_then(|s| ServerTimestamp::from_str(s).ok())
+        .ok_or_else(|| ErrorKind::MissingServerTimestamp)?)
+}
+
+/// A TimestampedResponse is used as a wrapper for any non-collection response
+/// we may want to update with an x-if-unmodified-since header to ensure we are
+/// overwriting what we think we are.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
```
Comment (linacambridge) [inline components/sync15/src/client.rs]: Is there a chance we might accidentally upload a `TimestampedResponse` as a BSO payload, instead of its record?

**Comment by mhammond** [comment]
subsumed by #892

**Pull request closed without merging**